src/SFA.DAS.EmployerAccounts.ReadStore/SFA.DAS.EmployerAccounts.ReadStore/Application/Commands/CreateAccountUserCommandHandler.cs

```diff
@@ -10,28 +12,38 @@ namespace SFA.DAS.EmployerAccounts.ReadStore.Application.Commands
     internal class CreateAccountUserCommandHandler : IReadStoreRequestHandler<CreateAccountUserCommand, Unit>
     {
         private readonly IAccountUsersRepository _accountUsersRepository;
+        private readonly ILogger _logger;
 
-        public CreateAccountUserCommandHandler(IAccountUsersRepository accountUsersRepository)
+        public CreateAccountUserCommandHandler(IAccountUsersRepository accountUsersRepository, ILogger logger)
         {
             _accountUsersRepository = accountUsersRepository;
+            _logger = logger;
         }
 
         public async Task<Unit> Handle(CreateAccountUserCommand request, CancellationToken cancellationToken)
         {
-            var user = await _accountUsersRepository.CreateQuery().SingleOrDefaultAsync(x => x.UserRef == request.UserRef && x.AccountId == request.AccountId, cancellationToken);
-
-            if (user == null)
+            try
             {
-                user = new AccountUser(request.UserRef, request.AccountId, request.Role, request.Created, request.MessageId);
-                await _accountUsersRepository.Add(user, null, cancellationToken);
+                var user = await _accountUsersRepository.CreateQuery().SingleOrDefaultAsync(x => x.UserRef == request.UserRef && x.AccountId == request.AccountId, cancellationToken);
+
+                if (user == null)
```

Comment:
We check if the user exists, and if it doesn't then assume that he user should be recreated.  what we don't check is if the user has already been recreated.

So the logic should be 

if(!UserExists)
{
   Create User
}
else if (UserAlreadyRecreated)
{
 Update Roles
}
else
{
 RecreateUser
}

Then the exception does not need to be caught.

Reply:
Ok I think I agree with the above

Reply:
so if (user == null) {
    CreateUser()
} else if (user.Deleted == null) {
   UpdateRoles()
} else if (user.Delete.HasValue) {
   RecreateUser();
}